### ISSUE TITLE
Support to reuse the byte array created by ZstdInputStream/ZstdOutputStream

### ISSUE THREAD

**Issue by chia7712** [luben/zstd-jni#131]
```java
    public ZstdInputStream(InputStream inStream) throws IOException {
        ...
        // allocate input buffer with max frame header size
        // no need to synchronize as these a finals
        this.src = new byte[srcBuffSize];
        ...
    }
```

The size of byte buffer is normally fixed so we can reuse the byte buffer after closing ```ZstdInputStream``` to avoid (expensive) allocation.

The approach is that we  add an new constructor to ZstdInputStream.
```java
public ZstdInputStream(InputStream inStream, Function<Integer, byte[]> bufferSupplier) {
  ...
  this.src = bufferSupplier.apply(srcBuffSize);
  ...
}
```

The proper type of ```bufferSupplier``` is ```java.util.function.Function```. However, it seems the target of project is JVM 1.6 so it is necessary to create an new class to offer the byte buffer.

Or the new constructor can accept byte array directly.
```java
public ZstdInputStream(InputStream inStream, byte[] buffer) {
  ...
  this.src = buffer
  ...
}
```
In order to create byte buffer with accurate size, we need to expose ```srcBuffSize```.
```
-    private static final int srcBuffSize = (int) recommendedDInSize();
+    public static final int srcBuffSize = (int) recommendedDInSize();
```

**Comment by chia7712** [comment]
I'd like to submit a PR for this issue :)

**Comment by luben** [comment]
Have you looked at the DirectByteBuffer variants - they are more flexible and will allow buffer reuse?

**Comment by chia7712** [comment]
@luben Thanks for response and pardon me that I did not give clear description.

The byte array I want to reuse is the field "src" of ```ZstdInputStream```. Our application produces a bunch of small streams. The default size of "src" ( 131075 bytes in our env) is a bit expensive when we create a lot of  ```ZstdInputStream``` instances. If we can control the allocation of byte buffer (src), we can reduce the memory consumption by reusing the byte array after closing the ```ZstdInputStream```

**Comment by luben** [comment]
Yes, I am open for PRs adding it, though I will be offline next 2 weeks and will look at it after I return.

**Comment by yuzawa-san** [comment]
I was looking over the snappy Java codebase and they use a BufferAllocatorFactory: https://github.com/xerial/snappy-java/blob/master/src/main/java/org/xerial/snappy/SnappyOutputStream.java#L94

What are your thoughts about using something similar? Especially in the ZstdOutputStream which also allocates a large byte[] in a similar manner.

Passing in a byte[] for internal usage seems like a code smell. However, if an BufferProvider or similar interface was used as the argument, then a default implementation could be included and folks could pass into their own implementation. I'd be willing to give it a try in a PR if you want.

Proposal: 
ZstdInputStream(InputStream) uses a default implementation to call ZstdInputStream(InputStream, BufferProvider)
ZstdOutputStream(OutputStream)  uses a default implementation to call ZstdOutputStream(OutputStream, BufferProvider)

BufferProvider would be something along the lines of https://github.com/xerial/snappy-java/blob/master/src/main/java/org/xerial/snappy/buffer/BufferAllocator.java

**Comment by chia7712** [comment]
@yuzawa-san Thanks for all your feedback.

I used ```byte[]``` rather than other supplier class since we use zstd-jni by dynamical reflection. That is related to license issue. Hence, the type of supplier must be a part of java standard library. The alternative way is to pass lambda function of JDK8. However, that demands this project update supported JDK from 6 to 8.

In short, the ```byte[]``` is good to apply zstd-jni by java reflection. Also, the byte array can be managed ourself.

> Proposal:
ZstdInputStream(InputStream) uses a default implementation to call ZstdInputStream(InputStream, BufferProvider)
ZstdOutputStream(OutputStream) uses a default implementation to call ZstdOutputStream(OutputStream, BufferProvider)

LGTM. That should be useful to users who can import zstd-jni directly.

**Comment by yuzawa-san** [comment]
@chia7712 So to be clear: the intent here is about reusing the buffer, rather than specifying a size? I see the test in your PR was updated to use multiple buffer sizes? However at the same time, you exposed the size of the srcBuffSize which implies an affinity to use that size only.

I'm putting together a proof of concept PR currently (I'll tag you on it of course) and I was playing around with actually completely hiding the recycling logic from the public API for this library. My rationale is from https://github.com/luben/zstd-jni/issues/141 where I was planning on adding buffer recycling and possibly an input buffer to the ZstdOutputStream as well. That is, I was going to hold off on the proposal I originally gave above and remove the new constructor with byte[] you added. What do you think of that? Does that work for your use case? If it about buffer reuse _and_ the buffer size then I think it would passing the buffer size as an int. This way the recycling can be hidden from the end user.

If not and you do want fine-grained control over the recycling logic, I could try my original proposal. In that case, would your reflection-based usage work with passing in a new BufferProvider implementation rather than passing a byte[]? You could implement within your own codebase a BufferProvider that returns byte[] of your own management. I suppose that I could try to provide an implementation bundled with this library if you are concerned about licensing and inclusion, but I would prefer leave that custom implementation up to the end-user.

**Comment by chia7712** [comment]
@yuzawa-san Thanks for your quickly response :)

> That is, I was going to hold off on the proposal I originally gave above and remove the new constructor with byte[] you added. What do you think of that? 

> In that case, would your reflection-based usage work with passing in a new BufferProvider implementation rather than passing a byte[]? You could implement within your own codebase a BufferProvider that returns byte[] of your own management.

My story is similar to your #141 but the trouble we encountered is that the memory pressure impacts the GC and performance. Hence, we have applied a patch (to reuse byte[] of compressed stream https://github.com/apache/kafka/pull/9220) to our kafka branch. Kafka uses zstd-jni by reflection so the patch I supplied to this issue is to use byte[] rather than supplier. If the BufferProvider you mentioned is from java standard library, it should be fine to me. Otherwise, it maybe a trouble to our use cases  (it is not easy to extend java class by reflection).

Could we keep both constructs (bytep[] and supplier) for ZstdOutputStream? It seems to me that they are not conflicting to each other.

**Comment by luben** [comment]
Hi, sorry for the delay on these issues - I was away for a few weeks.

@chia7712 , can you take a look at #143 - this approach seams good to me - it uses internal buffer poll so it will work with the reflection. Do you see any disadvantage for your use case?

Regards,
luben

**Comment by ijuma** [comment]
The current solution relies on soft references so the behavior is a bit unpredictable. For Apache Kafka, we would prefer to be able to do our own buffer management as we have to do it for lz4 anyway. Given the specifics of how we do decompression, we don't need any locking and or soft references.

**Comment by luben** [comment]
@ijuma : if we add another constructor variant that accepts a buffer (`byte[]`) so that you can wrap Zstd streams in an adapter that  handles you buffer pooling logic, is it going to help?

**Comment by ijuma** [comment]
@luben That could work. One question: what's the minimum Java version supported by zstd-jni?

**Comment by luben** [comment]
@ijuma, currently it's 1.6. That's why I am hesitant to add a lambda (buffer supplier) as another argument.

**Comment by ijuma** [comment]
Makes sense. Yes, wrapping should work. We would prefer `ByteBuffer` over `byte[]`, but the latter makes more sense for `ZstdInputStream`.

**Comment by luben** [comment]
@ijuma , there are `ZstdDirectBufferCompressingStream` and `ZstdDirectBufferDecompressingStream` that work on `DirectByteBuffer`s without any internal allocation. Downside is they don't implement Input/Output interfaces. Have you looked at them?

**Comment by yuzawa-san** [comment]
@ijuma can you link the buffer recycling implementation you wish to use, so we all can see what its interface is? Is it kafka's BufferSupplier. If we added a new interface in this repo, would it be feasible for the kafka project to provide their own implementation which could be a shim connected to BufferSupplier? Also do you imagine using the same recycler in the ZstdOutputStream. Can you elaborate on the reflection based instantiation? I'm guessing the dependency is optional and loaded on demand. Snappy seems to use their stock recycler (with the SoftReferences) https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/record/CompressionType.java#L90
Is the goal to overhaul that as well.

@luben what is the rationale for java 1.6 compatibility? it reached EOL in 2018. what are you thoughts about java 8 as a compromise?

**Comment by luben** [comment]
Yes, I know Java8 is EOL but there are some customers stuck on older versions. I still don't have plans to move it to Java 8 if it's not strictly necessary.

**Comment by chia7712** [comment]
Maybe it is time to reconsider to support both ways - byte[] and (global) pool. They are not in conflict with each other and that give more usages (control) to zstd users.

**Comment by luben** [comment]
@chia7712 , yes, this is what I am proposing. Though we have to do it both for the Input/Output streams. Can you take a look if you can modify your previous patch to add it?

**Comment by ijuma** [comment]
@yuzawa-san We have generally avoided direct dependencies on compression library classes due to the native dependencies that may not be available in certain systems. If it's only an interface, we can perhaps subclass that directly (avoiding the need for wrapping). @chia7712 Any concerns with that approach?

**Comment by chia7712** [comment]
> Any concerns with that approach?

If the interface of buffer supplier is not in java standard library, it may be a trouble to extend a java class by reflection. If zstd can support JDK 8, the java.util.Function is a good interface to be buffer supplier.

**Comment by ijuma** [comment]
@chia7712 We don't need to use reflection if we follow my suggestion.

**Comment by yuzawa-san** [comment]
@luben if we add some interfaces to be extended (like snappy does) that should not involve any loading of the native libraries right? If so, I think that is the best longer team solution. Kafka or any user can bring their own recycling implementations.

What are your thoughts about making the buffer recycler interface(s) using ByteBuffer versus byte[] and then adding the requirement that the ByteBuffer be from the heap. Then folks could even bring their own direct versus heap buffers with them if they so desired.

**Comment by luben** [comment]
@yuzawa-san , I am fine adding interface - subclassing it will not require loading native libraries.

I am a little bit unsure about making it work on ByteBuffers. Is there an easy way to get `byte[]` view into the ByteBuffer allocated memory? Else it will require reworking the jni C code that expects the arguments to be byte arrays.

**Comment by mortengrouleff** [comment]
The problem with on-heap ByteBuffers passed to jni code is that they block the rest of the JVM from doing GC. (See https://shipilev.net/jvm/anatomy-quarks/9-jni-critical-gclocker/)
So you should avoid passing such references to jni. If you want to use ByteBuffers, make them off-heap.

If you care for performance, why use the streams at all and not use the ZstdCompressCtx / ZstdDecompressCtx  instead?

**Comment by yuzawa-san** [comment]
@mortengrouleff yes, I've ran into GCLocker issues in the past with JNI. And I can confirm that the code as it stands today (with the off heap buffers, regardless of whether we use byte[] or HeapByteBuffer) is susceptible to GCLocker due to its usage of GetPrimitiveArrayCritical. I'm currently using this library via Kafka, which wraps InputStreams and OutputStreams for different compression libraries like so: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/record/CompressionType.java
 Perhaps @ijuma (as a prior contributor) may know more about the future of this and whether more bespoke or performance optimized implementations are underway. Given it appears the source and destinations in that case are both ByteBuffers (albeit on-heap) I guess maybe there are some improvements over using OutputStream and InputStream, but at the cost of interoperability between the libraries.

I'm putting together a POC of that BufferPool as argument which I believe is more elegant and safer than passing in your own byte[]. I'm working with the idea that Kafka can include its own implementation of BufferPool which wraps kafka's BufferSupplier. Given that their BufferSupplier uses HeapByteBuffers (and not byte[]), I'm planning on using ByteBuffer in the interface method signature. This makes the implementation cleaner. I was going to do something similar to https://github.com/apache/kafka/blob/b6ce9d68627bda904924ba2dce5c334a2c615e18/clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java#L79 and with that check for array (on heap). I was flirting with the idea of using direct buffers, but as @luben said that would require a larger overhaul, so I'm not pursuing that. If we did modify it such that the JNI used direct buffers, it would likely not be super worthwhile since we are implementing a FilterOutputStream/FilterInputStream which works in byte[]-land. So for ZstdOutputStream we would have to copy the source byte[] into a direct buffer run the compression and then take the destination direct buffer and copy that back over into another byte[] so we can pass to the wrapped OutputStream per the requirements of FilterOutputStream.

We can only dream that maybe future Java will have improved foreign function calls or memory pinning (panama?).

**Comment by yuzawa-san** [comment]
Here is a PR which exposes the new interface https://github.com/luben/zstd-jni/pull/146

**Comment by luben** [comment]
Merged. Closing as fixed

**Comment by luben** [comment]
@yuzawa-san , @chia7712 , I have released v1.4.5-10 that adds back the `throws IOException` to the `ZstdInputStream` constuctor: removing it introduced semver incompatibility.
